File: answers/lintcode/Valid Parentheses.cpp
class Solution {
public:
    /**
     * @param s A string
     * @return whether the string is a valid parentheses
     */
    bool isValidParentheses(string& s) {
        // Write your code here
		stack<char> parent_stack;
		for (size_t i = 0; i < s.length(); ++ i) {
			switch (s[i]) {
			case '(':
			case '[':
			case '{':
				parent_stack.push(s[i]);
				break;
			case ')':
				if (! parent_stack.empty() && '(' == parent_stack.top()) {
					parent_stack.pop();
				}
				else {
					return false;
				}
				break;
			case ']':
				if (! parent_stack.empty() && '[' == parent_stack.top()) {
					parent_stack.pop();
				}
				else {
					return false;
				}
				break;
			case '}':
				if (! parent_stack.empty() && '{' == parent_stack.top()) {
					parent_stack.pop();
				}
				else {
					return false;
				}
				break;
			}
		}
		return parent_stack.empty();
    }
};

File: answers/leetcode/Valid Parentheses/Valid Parentheses.java
class Solution {
    public boolean isValid(String s) {
        if (null == s) {
            return true;
        }
        
        Map<Character, Character> Parentheses = new HashMap<Character, Character>();
        Parentheses.put('(', ')');
        Parentheses.put('[', ']');
        Parentheses.put('{', '}');
        Stack<Character> stack = new Stack<Character>();
        for (char c : s.toCharArray()) {
            if (Parentheses.containsKey(c)) {
                stack.push(Parentheses.get(c));
            } else if (! stack.empty() && stack.peek() == c) {
                stack.pop();
            } else {
                return false;
            }
        }
        return stack.empty();
    }
}